File: solutions/3883-multiply-two-polynomials/multiply-two-polynomials.java
class Solution {
    // time = O((n + m) * log(n + m)), space = O(n + m)
    final double PI = Math.acos(-1);
    int[] rev;
    int bit, tot;
    public long[] multiply(int[] poly1, int[] poly2) {
        int n = poly1.length;
        int m = poly2.length;
        // find bit such that 2^bit >= n + m
        bit = 0;
        while ((1 << bit) < n + m) bit++;
        tot = 1 << bit;

        // dynamically allocate arrays of length tot
        Complex[] a = new Complex[tot];
        Complex[] b = new Complex[tot];
        rev = new int[tot];
        for (int i = 0; i < tot; i++) {
            a[i] = new Complex(0, 0);
            b[i] = new Complex(0, 0);
        }

        // copy input coefficients
        for (int i = 0; i < n; i++) a[i].x = poly1[i];
        for (int j = 0; j < m; j++) b[j].x = poly2[j];

        // build bit-reversal permutation
        for (int i = 0; i < tot; i++) {
            rev[i] = (rev[i >> 1] >> 1) | ((i & 1) << (bit - 1));
        }

        // forward FFT
        fft(a, 1);
        fft(b, 1);

        // point-wise multiply
        for (int i = 0; i < tot; i++) {
            a[i] = a[i].multiply(b[i]);
        }

        // inverse FFT
        fft(a, -1);

        // collect result and round
        long[] res = new long[n + m - 1];
        for (int i = 0; i < res.length; i++) {
            res[i] = Math.round(a[i].x / tot);
        }
        return res;
    }

    private void fft(Complex[] a, int inv) {
        // bit-reversal reorder
        for (int i = 0; i < tot; i++) {
            if (i < rev[i]) {
                Complex t = a[i];
                a[i] = a[rev[i]];
                a[rev[i]] = t;
            }
        }

        // Cooley–Tukey
        for (int len = 1; len < tot; len <<= 1) {
            double ang = PI / len * inv;
            Complex wlen = new Complex(Math.cos(ang), Math.sin(ang));
            for (int i = 0; i < tot; i += len << 1) {
                Complex w = new Complex(1, 0);
                for (int j = 0; j < len; j++) {
                    Complex u = a[i + j];
                    Complex v = a[i + j + len].multiply(w);
                    a[i + j] = u.plus(v);
                    a[i + j + len] = u.minus(v);
                    w = w.multiply(wlen);
                }
            }
        }
    }

    class Complex {
        private double x, y;
        public Complex(double x, double y) {
            this.x = x;
            this.y = y;
        }

        private Complex plus(Complex t) {
            return new Complex(x + t.x, y + t.y);
        }

        private Complex minus(Complex t) {
            return new Complex(x - t.x, y - t.y);
        }

        private Complex multiply(Complex t) {
            return new Complex(x * t.x - y * t.y, x * t.y + y * t.x);
        }
    }
}

File: solutions/3883-multiply-two-polynomials/multiply-two-polynomials.py
import numpy as np
from numpy.polynomial.polynomial import polymul

class Solution:
    def multiply(self, poly1: List[int], poly2: List[int]) -> List[int]:
        # Solution 1: WONT WORK
        # return np.rint(polymul(poly1, poly2)).astype(np.int64).tolist()

        # Solution 3: 
        return (lst := np.rint(polymul(poly1, poly2)).astype(np.int64).tolist()) \
       + [0] * ((len(poly1) + len(poly2) - 1) - len(lst))

        # Solution 2: 
        lst = np.rint(polymul(poly1, poly2)).astype(np.int64).tolist()
        n = len(poly1) + len(poly2) - 1
        if len(lst) < n:    
            lst.extend([0] * (n - len(lst)))
        return lst

        # Solution 4: 
        n1, n2 = len(poly1), len(poly2)
        n = n1 + n2 - 1
        N = 1 << (n - 1).bit_length() # next power of 2 ≥ n, efficient FFT
        fa = np.fft.rfft(poly1, n=N)  # real‐FFT of both padded polynomials
        fb = np.fft.rfft(poly2, n=N)
        fc = fa * fb                  # pointwise product in frequency domain
        res = np.fft.irfft(fc, n=N)   # invert and take real part
        res = np.rint(res[:n]).astype(np.int64) # round to int & make len n
        return res.tolist()

        # Solution 5: Slightly optimized solution 4
        n = len(poly1) + len(poly2) - 1
        N = 1 << (n - 1).bit_length()
        r = np.fft.irfft(np.fft.rfft(poly1, n=N)*np.fft.rfft(poly2, n=N),n=N)
        return np.rint(r[:n]).astype(int).tolist()


        # Solution 6:
        # bit‐reversal permutation + in‐place Cooley–Tuk FFT
        def fft(a: List[complex], invert: bool) -> None:
            n, j = len(a), 0        # bit reversal
            for i in range(1, n):
                bit = n >> 1
                while j & bit:
                    j ^= bit
                    bit >>= 1
                j |= bit
                if i < j: a[i], a[j] = a[j], a[i]

            length = 2
            while length <= n:      # Cooley–Tuk
                ang = 2 * math.pi / length * (-1 if invert else 1)
                wlen = complex(math.cos(ang), math.sin(ang))
                for i in range(0, n, length):
                    w = 1+0j
                    half = length >> 1
                    for k in range(i, i + half):
                        u = a[k]
                        v = a[k + half] * w
                        a[k] = u + v
                        a[k + half] = u - v
                        w *= wlen
                length <<= 1

            if invert:                  # divide by n on inverse
                for i in range(n):
                    a[i] /= n


        n1, n2 = len(poly1), len(poly2)
        n = n1 + n2 - 1                 # required result length
        N = 1                           # next power of two ≥ n
        while N < n: N <<= 1

        fa = list(map(complex, poly1)) + [0]*(N - n1) # zero‐pad
        fb = list(map(complex, poly2)) + [0]*(N - n2)
        fft(fa, False)                  # forward FFT
        fft(fb, False)
        for i in range(N):
            fa[i] *= fb[i]              # point‐wise multiplication
        fft(fa, True)                   # inverse FFT

        return [int(round(fa[i].real)) for i in range(n)]